### PR TITLE
io::Write::write_fmt: panic if the formatter fails when the stream does not fail

### PR DESCRIPTION
**Pull request by RalfJung**
Follow-up to https://github.com/rust-lang/rust/pull/124954

### CODE REVIEW

**Comment by RalfJung** [comment]
That test explicitly violates the existing std contract. It should probably be converted into a panic test?

File: tests/ui/write-fmt-errors.rs [hunk 1]
```diff
@@ -1,9 +1,11 @@
 //@ run-pass
+//@ needs-unwind
```
Comment (Nilstrieb) [inline tests/ui/write-fmt-errors.rs:2]: can you make it a should_panic unit test?

**Comment by Mark-Simulacrum** [comment]
r=me with or without unit test conversion

File: tests/ui/write-fmt-errors.rs (hunk 1 above)
Comment (RalfJung) [reply tests/ui/write-fmt-errors.rs:2]: Then I'd have to split it in 3 tests -- the parts that shouldn't panic, and the two panics.
I'd rather not do that.

**Comment by kpreid** [comment]
I think this *might* not be the right choice. IO errors, broadly, might happen for any operation for many reasons, and it feels “unhelpfully pedantic” to convert an error into a panic when there's no reason that it has to be a panic. The usual reasons to panic are

* there is no error return path available,
* recovering from the error would be incorrect, or
* it clearly indicates a bug *in the crate that is panicking or its caller*,

but this is none of those, because the panic is in `std`, but the bug could be in a crate that is neither `std` nor the crate that is invoking `io::Write::write_fmt()`. What about changing only the *message* to make it clear that this is a bug in the formatting routine?

On the other hand, this is straightforwardly a logic bug that is easy to not write (simply don't ever construct a `fmt::Error` unless you're implementing a `fmt::Write` stream).

**Comment by workingjubilee** [comment]
This seems like it is going to introduce non-elided panic branches into all the places where it is hardest to test for codegen. Is there a soundness-based motivation for the panic here?

**Comment by workingjubilee** [comment]
As for `format!`, it does not return an error, I believe.

**Comment by RalfJung** [comment]
> This seems like it is going to introduce non-elided panic branches into all the places where it is hardest to test for codegen.

I don't know what you mean by that.

> Is there a soundness-based motivation for the panic here?

No.

**Comment by RalfJung** [comment]
> I think this might not be the right choice. IO errors, broadly, might happen for any operation for many reasons, and it feels “unhelpfully pedantic” to convert an error into a panic when there's no reason that it has to be a panic.

This is not an IO error that's being converted here. Any such error in the formatter is a lingering panic anyway when used with `format!` (and likely with whatever form the "append format-args to String" operation will take).

So I would say it's helpfully pedantic. But I take your point, it is an avoidable panic.
We could tie it to the library UB checks... that's kind of a hack, it's not  UB check, but it would avoid the panic in release builds.

**Comment by workingjubilee** [comment]
@saethlin someone said your name /hj

**Comment by saethlin** [comment]
> This seems like it is going to introduce non-elided panic branches into all the places where it is hardest to test for codegen.

I also do not know what you mean by "hard to test for codegen". Looking at generated code is easy.

**Comment by workingjubilee** [comment]
@saethlin fair enough! my main thought was "but panicking involves formatting..." and that kind of problem.

**Comment by saethlin** [comment]
In general, the standard library formatting code tends to end up with a huge amount of indirection anyway. From a bottom-up optimization view I think the codegen is pretty much unsalvagable. The call graphs from this might be a bit goofy, but we're already doing panicky string slicing in the function that reports panics from string slicing, and even `Arguments::new` currently recurses: https://github.com/rust-lang/rust/pull/117804

I agree it's goofy and I'd like things to be different, but I think this PR adds some nice error detection capability which is probably the best sort of thing to do here.

**Comment by workingjubilee** [comment]
Alright, with a vote of "confidence" like that I'm fine with this, then, as long as we're also happy with letting this go in the future if it turns out to be a bad idea.

**Comment by RalfJung** [comment]
Yeah ofc, as with everything -- if it causes issues we can always revert. The joy of "undo". :)

@bors r=Mark-Simulacrum,workingjubilee

**Comment by RalfJung** [comment]
Uh, I did not expect this to have a perf impact on the compiler...

**Comment by Nilstrieb** [comment]
a tiny change on one benchmark, the benchmark doesn't seem to be very noisy recently from the graph but i don't think this is worth investigating.

**Comment by workingjubilee** [comment]
I'm more impressed on the 10% diff in RSS on regex (...but it's only like 30MB of a 300MB execution, I don't think it's worth worrying about).

**Comment by RalfJung** [comment]
RSS is very noisy, I have no idea where the threshold for significance is there.

**Pull request merged**